src/bootloaders/grub2.c

```diff
@@ -80,130 +100,284 @@ static inline char *grub2_get_boot_relative(void)
+                                                 tab,
                                                  BOOT_DIRECTORY, /* i.e. /boot */
                                                  kernel->target.initrd_path);
                 }
         }
 
         /* Finalize the entry */
-        cbm_writer_append(writer, "echo \"}\"\n");
+        cbm_writer_append_printf(config->writer, "echo \"%s}\"\n\n", root_tab);
 
-        cbm_writer_close(writer);
+        return true;
+}
 
-        conf_path = grub2_get_entry_path_for_kernel((BootManager *)manager, kernel);
+static bool grub2_write_config(const BootManager *manager, const Kernel *default_kernel)
+{
+        if (!manager) {
+                return false;
+        }
+
+        autofree(CbmWriter) *writer = CBM_WRITER_INIT;
+        autofree(char) *grub_dir = NULL;
+        const CbmDeviceProbe *root_dev = NULL;
+        const char *os_name = NULL;
+        const char *os_id = NULL;
+        autofree(char) *old_conf = NULL;
+        autofree(char) *conf_path = NULL;
+        autofree(char) *boot_dir = NULL;
+        const char *prefix = NULL;
+        bool is_separate;
+        Grub2Config config = { 0 };
+        bool wrote_submenu = false;
+
+        if (!cbm_writer_open(writer)) {
+                return false;
+        }
+
+        prefix = boot_manager_get_prefix((BootManager *)manager);
+        root_dev = boot_manager_get_root_device((BootManager *)manager);
+        if (!root_dev) {
+                LOG_FATAL("Root device unknown, this should never happen!");
+                return false;
+        }
+
+        is_separate = grub2_is_separate_boot_partition();
+        boot_dir = grub2_get_boot_relative();
+
+        os_name = boot_manager_get_os_name((BootManager *)manager);
+        os_id = boot_manager_get_os_id((BootManager *)manager);
+
+        /* Write out the stock header for our script */
+        cbm_writer_append(writer, "#!/bin/bash\nset -e\n");
+        cbm_writer_append(writer, ". \"/usr/share/grub/grub-mkconfig_lib\"\n");
+
+        /* Share our bits with grub2_write_kernel */
+        config = (Grub2Config){
+                .writer = writer,
+                .root_dev = root_dev,
+                .boot_dir = boot_dir,
+                .os_name = os_name,
+                .os_id = os_id,
+                .is_separate = is_separate,
+                .submenu = false,
+        };
+
+        /* Try to select a default kernel for update situations whereby CBM
+         * has been newly introduced, to ensure a /vmlinuz link
+         */
+        if (!default_kernel && kernel_queue->len == 1) {
+                default_kernel = nc_array_get(kernel_queue, 0);
+        }
+
+        /* Handle default kernel first always */
+        if (default_kernel) {
+                /* Attempt to clean out old files in migration, not fatal */
+                if (!grub2_remove_kernel(manager, default_kernel)) {
+                        LOG_ERROR("Failed to remove old kernel files for %s: %s",
```

Comment:
So this already logged (with a better filename) in grub2_remove_kernel. I guess the only additional information provided would be for the fact the remove was called here instead of somewhere else but I'm not sure how much it matters.